Update Analysis API to 2.0.20-dev-6911

vmishenev opened the pull request:
List of changes:
- KT-68462
- partially KT-68884 and others

whyoleg commented on dokka-subprojects/analysis-kotlin-symbols/src/main/kotlin/org/jetbrains/dokka/analysis/kotlin/symbols/services/SymbolFullClassHierarchyBuilder.kt:
```diff
@@ -127,21 +127,21 @@ internal class SymbolFullClassHierarchyBuilder(context: DokkaContext) : FullClas
 
         if (supersMap[typeConstructorWithKind.typeConstructor.dri] == null) {
-            val supertypes = kotlinType.getDirectSuperTypes(shouldApproximate = true).filterNot { it.isAny }
+            val supertypes = kotlinType.directSupertypes(shouldApproximate = true).filterNot { it.isAnyType }
```
the same here

whyoleg commented on dokka-subprojects/analysis-kotlin-symbols/src/main/kotlin/org/jetbrains/dokka/analysis/kotlin/symbols/services/SymbolFullClassHierarchyBuilder.kt:
```diff
@@ -127,21 +127,21 @@ internal class SymbolFullClassHierarchyBuilder(context: DokkaContext) : FullClas
 
     private fun KaSession.collectSupertypesWithKindFromKtType(
```
minor: `KtType` -> `KaType` in function name

whyoleg commented on dokka-subprojects/analysis-kotlin-symbols/src/main/kotlin/org/jetbrains/dokka/analysis/kotlin/symbols/translators/AnnotationTranslator.kt:
```diff
@@ -48,22 +48,22 @@ internal class AnnotationTranslator {
     }
 
-    private fun KaSession.mustBeDocumented(annotationApplication: KaAnnotationApplication): Boolean {
+    private fun KaSession.mustBeDocumented(annotationApplication: KaAnnotation): Boolean {
```
minor: `annotationApplication` -> `annotation`

whyoleg commented on dokka-subprojects/analysis-kotlin-symbols/src/main/kotlin/org/jetbrains/dokka/analysis/kotlin/symbols/translators/AnnotationTranslator.kt:
```diff
@@ -48,22 +48,22 @@ internal class AnnotationTranslator {
     }
 
-    private fun KaSession.toDokkaAnnotation(annotationApplication: KaAnnotationApplication) =
+    private fun KaSession.toDokkaAnnotation(annotationApplication: KaAnnotation) =
```
same

whyoleg commented on dokka-subprojects/analysis-kotlin-symbols/src/main/kotlin/org/jetbrains/dokka/analysis/kotlin/symbols/translators/DRIFactory.kt:
```diff
@@ -28,7 +28,7 @@ private fun CallableId.createDRI(receiver: TypeReference?, params: List<TypeRefe
 )
 
-internal fun getDRIFromNonErrorClassType(nonErrorClassType: KaNonErrorClassType): DRI =
+internal fun getDRIFromNonErrorClassType(nonErrorClassType: KaClassType): DRI =
```
minor: `getDRIFromNonErrorClassType` -> `getDRIFromKaClassType`, `nonErrorClassType` -> `classType`

whyoleg commented on dokka-subprojects/analysis-kotlin-symbols/src/main/kotlin/org/jetbrains/dokka/analysis/kotlin/symbols/translators/DRIFactory.kt:
```diff
@@ -61,13 +61,13 @@ internal fun KaSession.getDRIFromConstructor(symbol: KaConstructorSymbol): DRI {
 }
 
-internal fun KaSession.getDRIFromVariableLike(symbol: KaVariableLikeSymbol): DRI {
+internal fun KaSession.getDRIFromVariableLike(symbol: KaVariableSymbol): DRI {
```
same regarding renaming to match new naming

whyoleg commented on dokka-subprojects/analysis-kotlin-symbols/src/main/kotlin/org/jetbrains/dokka/analysis/kotlin/symbols/translators/DRIFactory.kt:
```diff
@@ -61,13 +61,13 @@ internal fun KaSession.getDRIFromConstructor(symbol: KaConstructorSymbol): DRI {
 }
 
-internal fun KaSession.getDRIFromFunctionLike(symbol: KaFunctionLikeSymbol): DRI {
+internal fun KaSession.getDRIFromFunctionLike(symbol: KaFunctionSymbol): DRI {
```
same

whyoleg commented on dokka-subprojects/analysis-kotlin-symbols/src/main/kotlin/org/jetbrains/dokka/analysis/kotlin/symbols/translators/DefaultSymbolToDocumentableTranslator.kt:
```diff
@@ -128,16 +127,16 @@ internal class DokkaSymbolVisitor(
-        val callables = scope.callables.toList().filterSymbolsInSourceSet(moduleFiles)
-        val classifiers = scope.classifiers.toList().filterSymbolsInSourceSet(moduleFiles)
+        val scope = packageSymbol.packageScope
+        val callables = scope.callables.filterSymbolsInSourceSet(moduleFiles)
```
`callables` and `classifiers` should be converted to `List` here, to avoid multiple recreation of sequences.

whyoleg commented on dokka-subprojects/analysis-kotlin-symbols/src/main/kotlin/org/jetbrains/dokka/analysis/kotlin/symbols/translators/DefaultSymbolToDocumentableTranslator.kt:
```diff
@@ -128,16 +127,16 @@ internal class DokkaSymbolVisitor(
 
-        val functions = callables.filterIsInstance<KaFunctionSymbol>().map { visitFunctionSymbol(it, dri) }
-        val properties = callables.filterIsInstance<KaPropertySymbol>().map { visitPropertySymbol(it, dri) }
+        val functions = callables.filterIsInstance<KaNamedFunctionSymbol>().map { visitFunctionSymbol(it, dri) }.toList()
```
question: is there any doc regarding the difference between `KaNamedFunctionSymbol` and `KaFunctionSymbol`?

vmishenev replied:
https://youtrack.jetbrains.com/issue/KT-68882/Analysis-API-Refactor-KaSymbols
> Rename the existing `KaFunctionSymbol` to KaNamedFunctionSymbol
> Rename `KaFunctionLikeSymbol` to KaFunctionSymbol

I think the hierarchy can answer:
![image](https://github.com/Kotlin/dokka/assets/9636937/7509ece9-b807-48d2-9c85-da2f9262274d)
Or what do you mean?

whyoleg commented on dokka-subprojects/analysis-kotlin-symbols/src/main/kotlin/org/jetbrains/dokka/analysis/kotlin/symbols/translators/DefaultSymbolToDocumentableTranslator.kt:
```diff
@@ -387,67 +386,70 @@ internal class DokkaSymbolVisitor(
         val constructors = scope.constructors.map { visitConstructorSymbol(it) }.toList()
 
-        val callables = scope.callables.toList()
+        val callables = scope.callables
```
same here about converting sequence to list, or we will recreate the sequence 3 times

whyoleg commented on dokka-subprojects/analysis-kotlin-symbols/src/main/kotlin/org/jetbrains/dokka/analysis/kotlin/symbols/translators/TypeTranslator.kt:
```diff
@@ -119,7 +119,7 @@ internal class TypeTranslator(
         type: KaType
     ): AncestryNode {
-        val (interfaces, superclass) = type.getDirectSuperTypes().filterNot { it.isAny }
+        val (interfaces, superclass) = type.directSupertypes(true).filterNot { it.isAnyType }
```
named parameter, also, AFAIU, old function used `false` here - is it expected that we change it?

vmishenev replied:
JIC it has no effect, but it might be more reliable in the case of unexpected `type`

whyoleg reviewed (changes requested):
Overall, nice!
Still, there are two things I want to resolve:
1. consistent naming: some declarations in AA changed their names, and so some of our functions are also changed, but others don't. I do have two suggestions here, it would be nice to consistently use one of them:
  1. we change everything to align with the new naming
  2. we change nothing until AA is stable enough and no renaming is expected
2. using of sequences: AFAIU AA almost all the time returns sequences, and in our case we need to handle most of the declarations from those sequences. I believe we should convert them to lists as soon as possible to not stick in a situation that we recreate those sequences on every call. This was the issue which I fixed not so long ago which causes big performance degradation in K2 analysis.

vmishenev commented:
>  consistent naming: some declarations in AA changed their names, and so some of our functions are also changed, but others don't. I do have two suggestions here, it would be nice to consistently use one of them:

    we change everything to align with the new naming
    we change nothing until AA is stable enough and no renaming is expected


I am trying to do the former to avoid forgotten context.
But renaming everything is quite hard since you should it manually. 
For example, I renamed only `namedClassOrObjectSymbol` to `namedClassSymbol` since here the name is crucial and scared for me.  

My approach is
 - rename only important changes;
 - after the AA stabilization, we can rename the remains if it makes sense

>  I believe we should convert them to lists as soon as possible to not stick in a situation that we recreate those sequences on every call.

I agree,  it is a good point. Using sequences from AA should be moderated.
I would say - in the end of a single chain.
